### ISSUE TITLE
implicit return types: do unreachable returns matter?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#20481]
The same idea (and proposed solution) as #14683 and #18597

What should the following function return?
```chpl
proc p() type {
  if true then
    return owned object;

  return borrowed object;
}
```

This is inspired by `proc borrowedToOwned()` in `test/classes/casts/class-type-cast2.chpl`

<details>

Today's behavior is consistent with [the spec](https://chapel-lang.org/docs/1.27/language/spec/procedures.html#implicit-return-types): there are two return statements and all return types can be implicitly converted to `borrowed object`. Even though the second return statement is unreachable, it participates in determining the return type.

Therefore the question in this issue is: is this the behavior that we want?

</details>

<details>
<summary>While there: we want a better error message</summary>

When resolving the following function:
```chpl
proc p() {
  if numLocales > 0 then
    return 1.2;
  else
    return "hi";
}
```
all I get is an error `unable to resolve return type` pointing at the line with `proc p()`.

I think we should clarify that the compiler was unable to find the best return type and list individual return statements and their types in the error message.

</detail>

**Comment by mppf** [comment]
@vasslitvinov --

Changing it to this makes the problem go away

``` chapel
proc p() type {
  if true then
    return owned object;
  else
    return borrowed object;
}

writeln(p():string);
```

As a result, I think this is actually a duplicate of issue #14683.

**Comment by vasslitvinov** [comment]
@mppf thank you, I did not see that issue.

Any thoughts on how to implement the solution?

**Comment by mppf** [comment]
AFAIK, it is just a matter of stopping the resolution process for a BlockStmt after a 'return'. I think this should be relatively straightforward in the dyno resolver.

**Comment by vasslitvinov** [comment]
Sounds good. How about:
```chpl
{ halt(); return xyz; }
```
versus
```chpl
{ someFunction(halt()); return xyz; }
```

where in the second case we do not want to resolve the call to someFunction either?

**Comment by vasslitvinov** [comment]
A perhaps more practical case:
```chpl
proc p() {
  if .... then return 5;
  else return 6;
  return "hi";     // should not be considered
}
```

**Comment by jhh67** [comment]
Here is my code that does not compile:
```
use Reflection;

record Foo {
    type t = int(32);
}

var foo = new Foo();

for param i in 0..<numFields(foo.type) {
    if isType(getField(foo,i)) {
        continue;
    }
    writeln(getField(foo, i));
}
```

Any statement that causes control to exit the loop at the point of the `continue` causes the same error.

**Comment by mppf** [comment]
@jhh67 - thanks for the example. Just making sure - you know that putting the `writeln` in an `else` should fix it, right?

**Comment by jhh67** [comment]
Yes, I was able to work around it.

**Comment by vasslitvinov** [comment]
I consider this resolved with #20497. Closing.
